File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/LocationOverlay.java
package net.cyclestreets.views.overlay;

import net.cyclestreets.util.Theme;
import net.cyclestreets.view.R;
import net.cyclestreets.views.CycleMapView;

import org.osmdroid.views.MapView;
import org.osmdroid.util.GeoPoint;
import org.osmdroid.views.overlay.mylocation.GpsMyLocationProvider;
import org.osmdroid.views.overlay.mylocation.MyLocationNewOverlay;

import android.app.Activity;
import android.content.Context;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Color;
import android.location.Location;
import android.location.LocationManager;
import android.support.design.widget.FloatingActionButton;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;

import static net.cyclestreets.util.MenuHelper.createMenuItem;
import static net.cyclestreets.util.MenuHelper.enableMenuItem;

public class LocationOverlay extends MyLocationNewOverlay {
  private final FloatingActionButton button_;

  private final CycleMapView mapView_;
  private final int onColor_;
  private final int followColor_;

  private boolean lockedOn_;

  private static class UseEverythingLocationProvider extends GpsMyLocationProvider {
    public UseEverythingLocationProvider(Context context) {
      super(context);
      LocationManager locMan = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
      for (String source : locMan.getProviders(true))
        addLocationSource(source);
    }
  }

  public LocationOverlay(final CycleMapView mapView) {
    super(new UseEverythingLocationProvider(mapView.getContext()), mapView.mapView());
    //setLocationUpdateMinTime(500);
    //setLocationUpdateMinDistance(10);

    mapView_ = mapView;

    onColor_ = Theme.lowlightColor(mapView_.getContext());
    followColor_ = Theme.highlightColor(mapView_.getContext()) | 0xFF000000;

    View overlayView = LayoutInflater.from(mapView_.getContext()).inflate(R.layout.locationbutton, null);
    button_ = overlayView.findViewById(R.id.locationbutton);
    button_.setOnClickListener(view -> enableAndFollowLocation(!isFollowLocationEnabled()));

    mapView_.addView(overlayView);

    lockedOn_ = false;
  }

  public void enableLocation(final boolean enable) {
    if (enable)
      enableMyLocation();
    else
      disableMyLocation();
  }

  public void enableAndFollowLocation(final boolean enable) {
    if (enable) {
      try {
        enableMyLocation();
        enableFollowLocation();
        final Location lastFix = getLastFix();
        if (lastFix != null)
          mapView_.getController().setCenter(new GeoPoint(lastFix));
      } catch (RuntimeException e) {
        // might not have location service
      }
    } else {
      disableFollowLocation();
      disableMyLocation();
    }

    mapView_.invalidate();
  }

  public void lockOnLocation() {
    lockedOn_ = true;
  }

  public void hideButton() {
    button_.setVisibility(View.INVISIBLE);
  }

  @Override
  public boolean onTouchEvent(final MotionEvent event, final MapView mapView) {
    final boolean handled = super.onTouchEvent(event, mapView);

    if (lockedOn_ && isMyLocationEnabled() && (event.getAction() == MotionEvent.ACTION_MOVE))
      enableFollowLocation();

    return handled;
  }

  ////////////////////////////////////////////
  @Override
  public void draw(final Canvas canvas, final MapView mapView, final boolean shadow) {
    if (isFollowLocationEnabled())
      button_.setColorFilter(followColor_);
    else
      button_.setColorFilter(isMyLocationEnabled() ? onColor_ : Color.TRANSPARENT);

    // I'm not thrilled about this but there isn't any other way (short of killing
    // and recreating the overlay) of turning off the little here-you-are man
    if (!isMyLocationEnabled())
      return;

    super.draw(canvas, mapView, shadow);
  }
}

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/LocationOverlay.kt
package net.cyclestreets.views.overlay

import android.Manifest.permission.ACCESS_FINE_LOCATION
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.location.LocationManager
import android.util.Log
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import androidx.core.content.res.ResourcesCompat
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.mikepenz.iconics.typeface.library.googlematerial.GoogleMaterial
import net.cyclestreets.iconics.IconicsHelper.materialIcon
import net.cyclestreets.util.Logging
import net.cyclestreets.util.Theme
import net.cyclestreets.util.doOrRequestPermission
import net.cyclestreets.view.R
import net.cyclestreets.views.CycleMapView
import org.osmdroid.util.GeoPoint
import org.osmdroid.views.MapView
import org.osmdroid.views.overlay.mylocation.GpsMyLocationProvider
import org.osmdroid.views.overlay.mylocation.MyLocationNewOverlay


private val TAG = Logging.getTag(LocationOverlay::class.java)


class LocationOverlay(private val mapView: CycleMapView) :
        MyLocationNewOverlay(UseEverythingLocationProvider(mapView.context), mapView.mapView()) {

    private val button: FloatingActionButton
    private val onColor: Int = Theme.lowlightColor(mapView.context)
    private val followColor: Int = Theme.highlightColor(mapView.context) or -0x1000000

    private class UseEverythingLocationProvider(context: Context) : GpsMyLocationProvider(context) {
        init {
            val locMan = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
            for (source in locMan.getProviders(true))
                addLocationSource(source)
        }
    }

    init {
        val context = mapView.context

        // Workaround for https://github.com/cyclestreets/android/issues/385 until the fix for
        // https://github.com/osmdroid/osmdroid/issues/1530 is available (in osmdroid 6.1.7 or above)
        // When the workaround is removed, the corresponding drawables should be also.
        val person = ResourcesCompat.getDrawable(context.resources, org.osmdroid.library.R.drawable.person, null) as BitmapDrawable
        val newDirectionArrow = ResourcesCompat.getDrawable(context.resources, R.drawable.temp_osmdroid_twotone_navigation_black_48, null) as BitmapDrawable
        setDirectionArrow(person.bitmap, newDirectionArrow.bitmap)
        // End workaround

        val overlayView = LayoutInflater.from(context).inflate(R.layout.locationbutton, null)
        button = overlayView.findViewById(R.id.locationbutton)
        button.setImageDrawable(materialIcon(context, GoogleMaterial.Icon.gmd_my_location))
        button.setOnClickListener { _ -> enableAndFollowLocation(!isFollowLocationEnabled) }

        mapView.addView(overlayView)
    }

    fun enableLocation(enable: Boolean) {
        if (enable)
            enableMyLocation()
        else
            disableMyLocation()
    }

    fun enableAndFollowLocation(enable: Boolean) {
        Log.d(TAG, "Location button clicked, enable = $enable")
        if (enable) {
            try {
                doOrRequestPermission(mapView.context, ACCESS_FINE_LOCATION) {
                    enableMyLocation()
                    enableFollowLocation()
                    val lastFix = lastFix
                    if (lastFix != null) {
                        Log.d(TAG, "Setting map centre to $lastFix")
                        mapView.controller.setCenter(GeoPoint(lastFix))
                    }
                }
            } catch (e: RuntimeException) {
                // might not have location service
                Log.d(TAG, "Failed to enable location finding", e)
            }
        } else {
            disableFollowLocation()
            disableMyLocation()
        }

        mapView.invalidate()
    }

    fun lockOnLocation() {
        setEnableAutoStop(false)
    }

    fun hideButton() {
        button.visibility = View.INVISIBLE
    }

    // Allows pinch-zoom while in LiveRide
    // see https://github.com/cyclestreets/android/issues/384
    // and https://github.com/osmdroid/osmdroid/issues/1578
    override fun onTouchEvent(event: MotionEvent, mapView: MapView?): Boolean {
        val isSingleFingerDrag = (event.action == MotionEvent.ACTION_MOVE)
                && (event.pointerCount == 1)

        if (event.action == MotionEvent.ACTION_DOWN && enableAutoStop) {
            disableFollowLocation()
        } else if (isSingleFingerDrag && isFollowLocationEnabled) {
            return true // prevent the pan
        }

        return false
    }

    ////////////////////////////////////////////
    override fun draw(canvas: Canvas, mapView: MapView, shadow: Boolean) {
        val buttonColor: Int
        if (isFollowLocationEnabled)
            buttonColor = followColor
        else if (isMyLocationEnabled)
            buttonColor = onColor
        else
            buttonColor = Color.TRANSPARENT

        button.setColorFilter(buttonColor)

        // I'm not thrilled about this but there isn't any other way (short of killing
        // and recreating the overlay) of turning off the little here-you-are man
        if (!isMyLocationEnabled)
            return

        super.draw(canvas, mapView, shadow)
    }
}
